File: backend/js_worker/src/dependencies.js
"use strict";

const fs = require("fs");
const path = require("path");
const acorn = require("acorn");
const walk = require("acorn-walk");

async function analyzeDependencies(files, repoRoot) {
  const internalPaths = new Set(files.map((f) => f.relPath));
  const edges = [];
  const nodeMap = new Map();

  const ensureNode = (moduleId, filePath, language, isExternal) => {
    if (!nodeMap.has(moduleId)) {
      nodeMap.set(moduleId, {
        module_id: moduleId,
        file_path: filePath || null,
        language,
        in_degree: 0,
        out_degree: 0,
        is_external: isExternal,
      });
    }
  };

  for (const file of files) {
    const fromMod = _pathToModuleId(file.relPath);
    ensureNode(fromMod, file.relPath, file.language, false);

    let source;
    try {
      source = fs.readFileSync(file.absPath, "utf8");
    } catch {
      continue;
    }

    let ast;
    try {
      ast = _parse(source, file.language);
    } catch {
      continue;
    }

    walk.simple(ast, {
      ImportDeclaration(node) {
        const toRaw = node.source.value;
        const { toMod, isExternal } = _resolveImport(toRaw, file.relPath, internalPaths);
        ensureNode(toMod, isExternal ? null : toMod + ".js", file.language, isExternal);
        edges.push({ from_module: fromMod, to_module: toMod, import_type: "import", is_external: isExternal });
      },
      CallExpression(node) {
        if (
          node.callee.name === "require" &&
          node.arguments.length === 1 &&
          node.arguments[0].type === "Literal"
        ) {
          const toRaw = node.arguments[0].value;
          const { toMod, isExternal } = _resolveImport(toRaw, file.relPath, internalPaths);
          ensureNode(toMod, isExternal ? null : toMod, file.language, isExternal);
          edges.push({ from_module: fromMod, to_module: toMod, import_type: "require", is_external: isExternal });
        }
      },
      // Dynamic import()
      ImportExpression(node) {
        if (node.source && node.source.type === "Literal") {
          const toRaw = node.source.value;
          const { toMod, isExternal } = _resolveImport(toRaw, file.relPath, internalPaths);
          ensureNode(toMod, null, file.language, isExternal);
          edges.push({ from_module: fromMod, to_module: toMod, import_type: "dynamic_import", is_external: isExternal });
        }
      },
    });
  }

  // Compute degrees
  for (const edge of edges) {
    if (nodeMap.has(edge.from_module)) nodeMap.get(edge.from_module).out_degree++;
    if (nodeMap.has(edge.to_module)) nodeMap.get(edge.to_module).in_degree++;
  }

  return { nodes: [...nodeMap.values()], edges };
}

function _pathToModuleId(relPath) {
  return relPath.replace(/\.(js|jsx|ts|tsx)$/, "").replace(/\\/g, "/");
}

function _resolveImport(importPath, fromRelPath, internalPaths) {
  // External package
  if (!importPath.startsWith(".") && !importPath.startsWith("/")) {
    return { toMod: importPath.split("/")[0], isExternal: true };
  }

  // Relative import
  const fromDir = path.dirname(fromRelPath);
  const resolved = path.join(fromDir, importPath).replace(/\\/g, "/");
  const withoutExt = resolved.replace(/\.(js|jsx|ts|tsx)$/, "");

  return { toMod: withoutExt, isExternal: false };
}

function _parse(source, language) {
  if (language === "typescript") {
    try {
      const tsParser = require("@typescript-eslint/typescript-estree");
      return tsParser.parse(source, { loc: true, range: true, jsx: true });
    } catch {}
  }
  return acorn.parse(source, {
    ecmaVersion: "latest",
    sourceType: "module",
    locations: true,
  });
}

module.exports = { analyzeDependencies };

File: backend/app/analyzers/python/dependencies.py
import ast
from pathlib import Path

from app.core.file_walker import FileInfo
from app.models.results import DependencyReport, DependencyEdge, DependencyNode

STDLIB_MODULES = frozenset([
    "os", "sys", "re", "json", "math", "io", "abc", "ast", "collections",
    "dataclasses", "datetime", "enum", "functools", "hashlib", "itertools",
    "logging", "pathlib", "pickle", "random", "shutil", "socket", "string",
    "subprocess", "tempfile", "threading", "time", "traceback", "typing",
    "unittest", "urllib", "uuid", "warnings",
])


def analyze_dependencies(files: list[FileInfo], repo_root: Path) -> DependencyReport:
    # Build a set of known internal module IDs for is_external detection
    internal_modules: set[str] = set()
    for fi in files:
        mod_id = _path_to_module_id(fi.rel_path)
        internal_modules.add(mod_id)

    edges: list[DependencyEdge] = []
    node_map: dict[str, DependencyNode] = {}

    for fi in files:
        from_mod = _path_to_module_id(fi.rel_path)
        _ensure_node(node_map, from_mod, fi.rel_path, fi.language, is_external=False)

        try:
            source = fi.path.read_text(encoding="utf-8", errors="ignore")
            tree = ast.parse(source, filename=fi.rel_path)
        except SyntaxError:
            continue

        for node in ast.walk(tree):
            if isinstance(node, ast.Import):
                for alias in node.names:
                    to_mod = alias.name.split(".")[0]
                    is_ext = _is_external(to_mod, internal_modules)
                    _ensure_node(node_map, alias.name, None, fi.language, is_external=is_ext)
                    edges.append(DependencyEdge(
                        from_module=from_mod,
                        to_module=alias.name,
                        import_type="import",
                        is_external=is_ext,
                    ))

            elif isinstance(node, ast.ImportFrom):
                if node.module is None:
                    continue
                base = node.module.split(".")[0]
                if node.level and node.level > 0:
                    # Relative import — resolve to sibling module
                    to_mod = _resolve_relative(fi.rel_path, node.module, node.level)
                    is_ext = False
                else:
                    to_mod = node.module
                    is_ext = _is_external(base, internal_modules)
                _ensure_node(node_map, to_mod, None, fi.language, is_external=is_ext)
                edges.append(DependencyEdge(
                    from_module=from_mod,
                    to_module=to_mod,
                    import_type="from_import",
                    is_external=is_ext,
                ))

    # Compute in/out degrees
    for edge in edges:
        if edge.from_module in node_map:
            node_map[edge.from_module].out_degree += 1
        if edge.to_module in node_map:
            node_map[edge.to_module].in_degree += 1

    return DependencyReport(nodes=list(node_map.values()), edges=edges)


def _path_to_module_id(rel_path: str) -> str:
    return rel_path.replace("/", ".").removesuffix(".py")


def _is_external(module_base: str, internal_modules: set[str]) -> bool:
    return module_base not in internal_modules


def _ensure_node(
    node_map: dict, module_id: str, file_path: str | None, language: str, is_external: bool
):
    if module_id not in node_map:
        node_map[module_id] = DependencyNode(
            module_id=module_id,
            file_path=file_path,
            language=language,
            is_external=is_external,
        )


def _resolve_relative(from_rel: str, module: str, level: int) -> str:
    parts = from_rel.split("/")
    # Go up `level` directories
    base_parts = parts[:-(level)]
    if module:
        return ".".join(base_parts + [module])
    return ".".join(base_parts)
